sdk/core/Azure.Core.Amqp/src/AmqpMessageBody.cs

```diff
@@ -121,5 +151,40 @@ public bool TryGetSequence(out IEnumerable<IList<object>>? sequence)
             sequence = null;
             return false;
         }
+
+        private static void ValidateAmqpPrimitive(object value, string paramName)
+        {
+            Argument.AssertNotNull(value, paramName);
+            switch (value)
+            {
+                case string:
+                case byte:
+                case sbyte:
+                case char:
+                case short:
+                case ushort:
+                case int:
+                case uint:
+                case long:
+                case ulong:
+                case float:
+                case double:
+                case decimal:
+                case bool:
+                case Guid:
+                case DateTime:
+                    return;
+                case IEnumerable enumerable:
```

Comment:
Since the `IEnumerable` is passed in, we shouldn't consider it safe to enumerate multiple times.  Given the potential for nested types, this would be really tricky to unroll into a list or other known-safe collection type, though.

Maybe the best that we're able to reasonably do would be to explicitly document any place where a message body is created (explicitly or implicitly) that we require that any enumerable types be safe to enumerate multiple times.